File: 226.invert-binary-tree.py
#
# @lc app=leetcode id=226 lang=python3
#
# [226] Invert Binary Tree
#

# Time:  O(n)
# Space: O(h)

import collections


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if not root:
            return root

        # iterative
        queue = collections.deque([root])
        while queue:
            node = queue.popleft()
            node.left, node.right = node.right, node.left
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        return root

        # recursive
        root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
        return root


# @lc code=end

File: 226.invert-binary-tree.js
/*
 * @lc app=leetcode id=226 lang=javascript
 *
 * [226] Invert Binary Tree
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function (root) {
  if (!root) {
    return root;
  }
  [root.left, root.right] = [root.right, root.left];
  root.left = invertTree(root.left);
  root.right = invertTree(root.right);
  return root;
};
// @lc code=end
